src/modm/platform/fault/crashcatcher/crashcatcher_handler.cpp.in

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2016, 2018, Niklas Hauser
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#include <modm/platform/uart/uart_{{ id }}.hpp>
+#include <CrashCatcher.h>
+
+extern "C" modm_weak bool
+modm_hardfault_coredump_start()
+{ return true; }
+
+extern "C" const CrashCatcherMemoryRegion*
+CrashCatcher_GetMemoryRegions(void)
+{
+	static const CrashCatcherMemoryRegion regions[] = {
+%% for m in memory
+		{ {{ m["start"] }}, {{m["start"]}} + {{m["size"]}}, CRASH_CATCHER_BYTE }, // {{ m["name"] | upper }}
+%% endfor
+		{ 0xFFFFFFFF, 0xFFFFFFFF, CRASH_CATCHER_BYTE }
+	};
+	return regions;
+}
+
+MODM_ISR_DECL({{ uart | upper ~ id }});
+using Uart = modm::platform::{{ uart ~ id }};
+using UartHal = modm::platform::{{ uart }}Hal{{ id }};
+// since the hard fault handler cannot be preempted by any other handler or IRQ
+// we have to manually call it here. Not pretty, but it works.
+void flushUart()
+{
+	while(not Uart::isWriteFinished())
+	{
+		if (UartHal::isTransmitRegisterEmpty())
+		{
+			MODM_ISR_CALL({{ uart | upper ~ id }});
+		}
+	}
+}
+
+void printStr(const char *str)
+{
+	while(*str)
+		Uart::write(*str++);
+	flushUart();
+}
+
+extern "C" void
+CrashCatcher_putc(int c)
+{ Uart::write(c); flushUart(); }
+
+extern "C" int
+CrashCatcher_getc(void)
+{ return -1; }
+
+extern "C" void
+CrashCatcher_DumpStart(const CrashCatcherInfo *)
+{
+	if (not modm_hardfault_coredump_start())
+		NVIC_SystemReset();
+	Uart::discardTransmitBuffer();
+	printStr("\n\n>>> HARDFAULT\n\n");
+}
+
+void hexlify(uint8_t byte)
+{
+	CrashCatcher_putc( "0123456789ABCDEF"[byte >> 4] );
+	CrashCatcher_putc( "0123456789ABCDEF"[byte & 0xF] );
+}
+
+void dumpMemory(const uint8_t *buffer, size_t size)
+{
+	for (size_t ii = 1; ii <= size; ii++)
+	{
+		// 16 bytes => line break
+		if (ii % 16 == 0) CrashCatcher_putc('\n');
+		hexlify(*buffer++);
+	}
+}
+
+extern "C" void
+CrashCatcher_DumpMemory(const void* pvMemory,
+						CrashCatcherElementSizes elementSize,
+						size_t elementCount)
+{
+	switch (elementSize)
+	{
+		case CRASH_CATCHER_BYTE:
+			dumpMemory((const uint8_t *) pvMemory, elementCount);
+			break;
+		case CRASH_CATCHER_HALFWORD:
+		{
+			const uint16_t *pMemory = (const uint16_t *) pvMemory;
+			for (size_t ii = 1; ii <= elementCount; ii++)
+			{
+				uint16_t val = *pMemory++;
```

Comment:
To be pedantic, dereferencing pMemory here is undefined behaviour. According to language rules commonly referred to as "strict aliasing rules" pMemory has to point to an actual (u)int16_t object to be safe to dereference (https://en.cppreference.com/w/cpp/language/reinterpret_cast#Type_aliasing). Although every sensible compiler will probably generate the expected code, this pitfall is easy to avoid here. There is an exception to these rules for (unsigned) char and std::byte which may alias any type. If I am not overlooking something, it is possible to define pMemory as uint8_t* and to pass it to dumpMemory directly.

Reply:
> If I am not overlooking something, it is possible to define pMemory as uint8_t* and to pass it to dumpMemory directly.

Some memories can only be read with 16- or 32-bit accesses (mostly the peripherals), hence this annoyingly duplicate code. I can't change the `const void *pvMemory`, since that's the callback API from CrashCatcher, but how would you do this correctly?

Reply:
> how would you do this correctly?

To me it does not make sense at all, to apply these rules to memory-mapped IO. The rules deal with accessing C++ objects in memory through a pointer of a different type. Applying the concept of a C++ object to some piece of hardware is totally nonsensical. There is no existing C++ object in memory that the compiler knows of. Hence, the only practical solution is to leave the code as it is. 

However I don't think there is any guarantee that the intended 16-bit access will actually be performed as one. I haven't looked much into the assembly output of arm compilers but on modern x86 platforms simple byte copying operations in loops often turn into fancy AVX instructions.
Using a pointer to volatile for the read would clearly state to the compiler: I want a 16 bit read and I want it now. I suppose adding volatile would not impede wanted optimizations. As the pointer itself is not volatile but the data it points to, operations on the pointer will still be optimized. I compiled a few toy examples for cortex-m4 with -Os and the resulting assembly looked almost identical regardless of whether volatile is added or not.

Reply:
Good observation!
Another way to get around UB could be to add a `read16` function coded in assembly that includes the correct load instruction.

Reply:
Added it to the TODO list in the new PR.